src/lib/core/datetime.c

```diff
@@ -861,3 +865,187 @@ interval_interval_add(struct interval *lhs, const struct interval *rhs)
 }
+
+/** This structure contains information about the given date and time fields. */
+struct dt_fields {
+	/* Specified year. */
+	double year;
+	/* Specified month. */
+	double month;
+	/* Specified day. */
+	double day;
+	/* Specified hour. */
+	double hour;
+	/* Specified minute. */
+	double min;
+	/* Specified second. */
+	double sec;
+	/* Specified millisecond. */
+	double msec;
+	/* Specified microsecond. */
+	double usec;
+	/* Specified nanosecond. */
+	double nsec;
+	/* Specified timestamp. */
+	double timestamp;
+	/* Specified timezone offset. */
+	double tzoffset;
+	/* Number of given fields among msec, usec and nsec. */
+	int count_usec;
+	/* True, if any of year, month, day, hour, min or sec is specified. */
+	bool is_ymdhms;
+	/* True, if timestamp is specified. */
+	bool is_ts;
+};
+
+/** Parse msgpack value and convert it to double, if possible. */
+static int
+get_double_from_mp(const char **data, double *value)
+{
+	switch (mp_typeof(**data)) {
+	case MP_INT:
+		*value = mp_decode_int(data);
+		break;
+	case MP_UINT:
+		*value = mp_decode_uint(data);
+		break;
+	case MP_DOUBLE:
+		*value = mp_decode_double(data);
+		break;
+	case MP_EXT: {
+		int8_t type;
+		uint32_t len = mp_decode_extl(data, &type);
+		if (type != MP_DECIMAL)
+			return -1;
+		decimal_t dec;
+		if (decimal_unpack(data, len, &dec) == NULL)
+			return -1;
+		*value = atof(decimal_str(&dec));
+		break;
+	}
+	default:
+		return -1;
+	}
+	return 0;
+}
+
+/** Define field of DATETIME value from field of given MAP value.*/
+static int
+map_field_to_dt_field(struct dt_fields *fields, const char **data)
+{
+	if (mp_typeof(**data) != MP_STR) {
+		mp_next(data);
+		mp_next(data);
+		return 0;
+	}
+	uint32_t size;
+	const char *str = mp_decode_str(data, &size);
+	double *value;
+	if (strncmp(str, "year", size) == 0) {
+		value = &fields->year;
+		fields->is_ymdhms = true;
+	} else if (strncmp(str, "month", size) == 0) {
+		value = &fields->month;
+		fields->is_ymdhms = true;
+	} else if (strncmp(str, "day", size) == 0) {
+		value = &fields->day;
+		fields->is_ymdhms = true;
+	} else if (strncmp(str, "hour", size) == 0) {
+		value = &fields->hour;
+		fields->is_ymdhms = true;
+	} else if (strncmp(str, "min", size) == 0) {
+		value = &fields->min;
+		fields->is_ymdhms = true;
+	} else if (strncmp(str, "sec", size) == 0) {
+		value = &fields->sec;
+		fields->is_ymdhms = true;
+	} else if (strncmp(str, "msec", size) == 0) {
+		value = &fields->msec;
+		++fields->count_usec;
+	} else if (strncmp(str, "usec", size) == 0) {
+		value = &fields->usec;
+		++fields->count_usec;
+	} else if (strncmp(str, "nsec", size) == 0) {
+		value = &fields->nsec;
+		++fields->count_usec;
+	} else if (strncmp(str, "timestamp", size) == 0) {
+		value = &fields->timestamp;
+		fields->is_ts = true;
+	} else if (strncmp(str, "tzoffset", size) == 0) {
+		value = &fields->tzoffset;
+	} else {
+		mp_next(data);
+		return 0;
+	}
+	return get_double_from_mp(data, value);
```

Comment:
There is one field in the table that is accepted by `require('datetime').new()` that correctly accepts fractional numbers values, and that is the `timestamp` field. In addition, all other fields in this function accept fractional numbers (note that `require('datetime').interval.new()` does not accept fractional numbers). Although for all fields except `timestamp`, this number is simply truncated. To mimic this behavior, I decided to use `double` instead of `int64_t` or `int32_t`.